packages/server/src/common/command/command-stack.ts

```diff
@@ -13,62 +13,131 @@
  */
 export interface CommandStack {
     /**
-     * Clears any redoable commands not yet redone, adds the command, and then executes the command.
+     * Clears any redoable commands not yet redone, adds the command to the stack and then invokes {@link Command.execute}.
      * @param command the command to execute.
      */
     execute(command: Command): void;
 
     /**
-     * Moves the top of the stack down, undoing what was formerly the top command.
+     * Removes the topmost (i.e. last executed) command from the stack and invokes {@link Command.undo}.
      */
     undo(): void;
 
     /**
-     * Moves the top of the stack up, redoing the new top command.
+     * Returns `true` if the top command on the stack can be undone.
+     */
+    canUndo(): boolean;
+
+    /**
+     * Re-adds the last undo command on top of the stack and invokes {@link Command.redo}
      */
     redo(): void;
 
     /**
-     * The command stack has currently unsaved changes.
+     * Returns `true` if there are  redoable commands in the stack.
      */
-    isDirty(): boolean;
+    canRedo(): boolean;
 
     /**
      * Called after a save has been successfully performed.
      */
     saveIsDone(): void;
+
+    /**
+     * Indicates wether the command stack has unsaved changes.
+     */
+    readonly isDirty: boolean;
+
+    /**
+     * Disposes all the commands in the stack.
+     */
+    flush(): void;
 }
 
 @injectable()
 export class DefaultCommandStack implements CommandStack {
+    @inject(Logger)
+    protected logger: Logger;
+
+    protected undoStack: Command[] = [];
+    protected redoStack: Command[] = [];
+
     execute(command: Command): void {
-        // no-op
+        this.redoStack = [];
+        try {
+            command.execute();
+        } catch (error) {
+            this.handleError(error);
+        }
+        this.undoStack.push(command);
     }
 
     undo(): void {
-        // no-op
+        const command = this.undoStack.pop();
+        if (command && Command.canUndo(command)) {
+            try {
+                command.undo();
+            } catch (error) {
+                this.handleError(error);
+            }
+            this.redoStack.push(command);
+        }
+    }
+
+    canUndo(): boolean {
+        return this.undoStack.length > 0 //
+            ? Command.canUndo(this.undoStack[this.undoStack.length - 1])
+            : false;
     }
 
     redo(): void {
-        // no-op
+        const command = this.redoStack.pop();
+        if (command) {
+            try {
+                command.redo();
+            } catch (error) {
+                this.handleError(error);
+            }
+            this.undoStack.push(command);
+        }
+    }
+
+    canRedo(): boolean {
+        return this.redoStack.length > 0;
     }
 
-    isDirty(): boolean {
-        return false;
+    get isDirty(): boolean {
+        return this.undoStack.length > 0;
     }
 
     saveIsDone(): void {
-        // no-op
+        this.flush();
```

Comment:
This behavior does not match the Java side as with the `flush` we can no longer do any `undo` after the save which is a bit unexpected. In Java/EMF, the behavior is to simply remember the last save index and compare it to the index of the last executed operation, i.e., it is more a shifting of pointers in the command history. I think it would be nice to match this behavior here as well, what do you think?

Reply:
Good catch. I have updated to commandstack implementation to be inline with the EMF implementation (41f9759)